pandas/hashtable.pyx

```diff
@@ -80,72 +61,120 @@ cdef class ObjectVector:
         self.data[self.n] = <PyObject*> o
         self.n += 1
 
+    def to_array(self):
+        self.ao.resize(self.n)
+        self.m = self.n
+        return self.ao
+
+ctypedef struct Int64VectorData:
+    int64_t *data
+    size_t n, m
+
+ctypedef struct Float64VectorData:
+    float64_t *data
+    size_t n, m
+
+ctypedef fused vector_data:
+    Int64VectorData
+    Float64VectorData
+
+ctypedef fused sixty_four_bit_scalar:
+    int64_t
+    float64_t
+
+cdef bint needs_resize(vector_data *data) nogil:
+    return data.n == data.m
+
+cdef void append_data(vector_data *data, sixty_four_bit_scalar x) nogil:
+
+    # compile time specilization of the fused types
+    # as the cross-product is generated, but we cannot assign float->int
+    # the types that don't pass are pruned
+    if (vector_data is Int64VectorData and sixty_four_bit_scalar is int64_t) or (
```

Comment:
@cpcloud I got this to work after puzzling on what cython does with multiple fused types. It creates the cross-product, but you generally need only certain specilizations; it does this beaut as a compile-time prune, so no perf hit.